src/seq_analysis/orf.rs

```diff
@@ -0,0 +1,161 @@
+                                          x.into_iter()                     // Vec<&[u8;3]> to
+                                           .map(|&x| x as u8)               // Vec<VecDeque<u8>>
+                                           .collect::<VecDeque<u8>>()       // so they can be
+                                      })                                    // easily compared
+                                      .collect(), // with codon built
+            stop_codons: stop_codons.into_iter()                            // from IntoTextIterator
+                                    .map(|x| {                              // object.
+                                        x.into_iter()
+                                         .map(|&x| x as u8)
+                                         .collect::<VecDeque<u8>>()
+                                    })
+                                    .collect(),
+            min_len: min_len,
+        }
+    }
+
+    /// Find all orfs in the given sequence
+    pub fn find_all<'a, I: IntoTextIterator<'a>>(&'a self, seq: I) -> Matches<I::IntoIter> {
+        Matches {
+            finder: self,
+            state: State::new(),
+            seq: seq.into_iter().enumerate(),
+        }
+    }
+}
+
+
+/// The current algorithm state.
+struct State {
+    start_pos: [usize; 3],
+    in_orf: [bool; 3],
+    orf: [Vec<u8>; 3],
+    codon: VecDeque<u8>,
+}
+
+
+impl State {
+    /// Create new state.
+    pub fn new() -> Self {
+        State {
+            start_pos: [0, 0, 0],
+            in_orf: [false, false, false],
+            orf: [Vec::new(), Vec::new(), Vec::new()],
+            codon: VecDeque::new(),
+        }
+    }
+}
+
+
+/// Iterator over offset, start position, end position and sequence of matched orfs.
+pub struct Matches<'a, I: TextIterator<'a>> {
+    finder: &'a Finder,
+    state: State,
+    seq: iter::Enumerate<I>,
+}
+
+impl<'a, I: Iterator<Item = &'a u8>> Iterator for Matches<'a, I> {
+    type Item = (usize, usize, usize, Vec<u8>);
+
+    fn next(&mut self) -> Option<(usize, usize, usize, Vec<u8>)> {
+
+        let mut result: Option<(usize, usize, usize, Vec<u8>)> = None;
+        let mut offset: usize;
+
+        for (index, &nuc) in self.seq.by_ref() {
+
+            // update the codon
+            if self.state.codon.len() >= 3 {
+                self.state.codon.pop_front();
+            }
+            self.state.codon.push_back(nuc);
+            offset = (index + 1) % 3;
+
+            // inside orf
+            if self.state.in_orf[offset] {
+                // check if should stop
+                if self.finder.stop_codons.contains(&self.state.codon) {
+                    // add last codon
+                    self.state.orf[offset].extend(self.state.codon.clone().into_iter());
+                    // build results
+                    if self.state.orf[offset].len() > self.finder.min_len {
+                        result = Some((offset,
```

Comment:
Why can't we just return an interval with start and stop (exclusive) in the given sequence. This way, we don't need the allocation for the orf sequence. If people want the orf sequence, they can slice the original sequence with the returned interval.